src/Raven.Studio/typescript/common/queryCompleter.ts

```diff
@@ -0,0 +1,279 @@
+        this.collectionsTracker = collectionsTracker.default;
+    }
+    
+    private getIndexName(session: AceAjax.IEditSession): [string, boolean] {
+        let keyword: string;
+        
+        for (let row = 0; row < session.getLength(); row++) {
+            let lineTokens: AceAjax.TokenInfo[] = session.getTokens(row);
+
+            for (let i = 0; i < lineTokens.length; i++) {
+                const token = lineTokens[i];
+                switch (token.type) {
+                    case "keyword": {
+                        keyword = token.value.toLowerCase();
+                        break;
+                    }
+                    case "string": {
+                        const indexName = token.value.substr(1, token.value.length - 2);
+                        if (keyword === "from")
+                            return [indexName, false];
+                        if (keyword === "index")
+                            return [indexName, true];
+                        break;
+                    }
+                    case "identifier": {
+                        const indexName = token.value;
+                        if (keyword === "from")
+                            return [token.value, false];
+                        if (keyword === "index")
+                            return [indexName, true];
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    private getIndexFields(session: AceAjax.IEditSession): JQueryPromise<string[]> {
+        
+        const [indexName, isStaticIndex] = this.getIndexName(session);
+        if (!indexName) {
+            return $.when<string[]>([]);
+        }
+
+        const cache = this.indexFieldsCache.get(indexName);
+        if (cache) {
+            return $.when<string[]>(cache);
+        }
+
+        if (isStaticIndex) {
+            return new getIndexEntriesFieldsCommand(indexName, this.activeDatabase())
+                .execute()
+                .then((fields) => {
+                    this.indexFieldsCache.set(indexName, fields.Results);
+                    return $.when(fields.Results);
+                });
+        } else {
+            return new collection(indexName, this.activeDatabase())
+                .fetchDocuments(0, 1)
+                .then(result => {
+                    // TODO: Modify the command to return also nested pathes, like Address.City
+                    if (result && result.items.length > 0) {
+                        const propertyNames = new document(result.items[0]).getDocumentPropertyNames();
+                        this.indexFieldsCache.set(indexName, propertyNames);
+                        return $.when(propertyNames);
+                    }
+                    return $.when<string[]>([]);
+                });
+        }
+    }
+
+    private getLastKeyword(session: AceAjax.IEditSession, pos: AceAjax.Position): [string, string, string, string, number] {
```

Comment:
this function can't remove: [string, string, string, string, number]
I think we should introduce returned type as object:
{ keyword: string, operator: string, ....}